service/history/replicationTaskProcessor.go

```
@@ -0,0 +1,285 @@
+				p.logger.Info("Closing replication task processor.", tag.ShardID(p.shard.GetShardID()), tag.ReadLevel(p.readLevel))
+				return
+			}
+		}
+	}()
+
+	p.logger.Info("ReplicationTaskProcessor started.")
+}
+
+func (p *replicationTaskProcessor) Stop() {
+	close(p.done)
+}
+
+func (p *replicationTaskProcessor) processTask(replicationTask *r.ReplicationTask) {
+	var err error
+SubmitLoop:
+	for {
+		var scope int
+		switch replicationTask.GetTaskType() {
+		case r.ReplicationTaskTypeDomain:
+			scope = metrics.DomainReplicationTaskScope
+			err = p.handleDomainReplicationTask(replicationTask)
+		case r.ReplicationTaskTypeSyncShardStatus:
+			scope = metrics.SyncShardTaskScope
+			err = p.handleSyncShardTask(replicationTask)
+		case r.ReplicationTaskTypeSyncActivity:
+			scope = metrics.SyncActivityTaskScope
+			err = p.handleActivityTask(replicationTask)
+		case r.ReplicationTaskTypeHistory:
+			scope = metrics.HistoryReplicationTaskScope
+			err = p.handleHistoryReplicationTask(replicationTask)
+		case r.ReplicationTaskTypeHistoryMetadata:
+			// Without kafka we should not have size limits so we don't necessary need this in the new replication scheme.
+		default:
+			p.logger.Error("Unknown task type.")
+			scope = metrics.ReplicatorScope
+			err = ErrUnknownReplicationTask
+		}
+
+		if err != nil {
+			p.logger.Info("Failed to apply replication task.", tag.TaskID(replicationTask.GetSourceTaskId()))
+			p.updateFailureMetric(scope, err)
+			if !isTransientRetryableError(err) {
+				p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster)).IncCounter(metrics.ReplicationTasksFailed)
+				break SubmitLoop
+			}
+		} else {
+			p.logger.Info("Successfully applied replication task.", tag.TaskID(replicationTask.GetSourceTaskId()))
+			p.metricsClient.Scope(metrics.ReplicationTaskFetcherScope, metrics.TargetClusterTag(p.sourceCluster)).IncCounter(metrics.ReplicationTasksApplied)
+			break SubmitLoop
+		}
+	}
+
+	if err != nil {
```

I think you may want to return an error to the caller.

---

Caller cannot do anything in this case. We can at best put in our own DLQ and then it will require human intervention.